source/common/config/utility.h

```diff
@@ -292,9 +292,12 @@ class Utility {
    */
   template <class Factory, class ProtoMessage>
   static Factory* getFactory(const ProtoMessage& message) {
-    Factory* factory = Utility::getFactoryByType<Factory>(message.typed_config());
-    if (factory != nullptr) {
-      return factory;
+    if (message.has_typed_config()) {
```

Comment:
This seems like a step in the wrong direction.  My understanding is that we want to go in the direction that the type_url always defines the extension to use.  It should not be optional -- we should reject extensions for which there is no embedded config type.  If an extension does not need any config, it should still define its own config message type, which can be empty.  That way, we still have a way to uniquely identify it.

Note that gRPC already does the above.  It has no concept of Envoy's extension names; the extension name idea is considered legacy and IMHO should be removed from Envoy.

Reply:
The problem is the legacy definitions that allow empty typed_config (e.g. HTTP filters in listeners). I think the new definitions already require defined typed_config in which case this condition is always true, but we still have instances without any typed_config in the wild and in the tests here. The proposal is basically restrict the usage of extension names only to those remaining instances without any typed_config.

Reply:
My concern is that this change does nothing to prevent new extensions from being added without a typed_config.  And by documenting the fact that this does work, it actually seems to encourage people into doing that in the future, on the incorrect assumption that this is a supported and intended approach.

I would really prefer to lock this down rather than documenting the fact that it works.  How about the following instead:

- Fix all remaining extensions in the Envoy repo to have typed_configs.
- Add a runtime option to reject extensions that do not specify a typed_config.  Initially, this will be off by default, but in release N+1, it will be changed to be on by default.
- Add a log message for any extension that does not specify a typed_config, warning people that the config will stop working by default in release N+1.
- At some point in the future, we can remove the runtime option and support only extensions that specify typed_config.

@mattklein123 or @htuch, thoughts?

Reply:
@markdroth The specific problem that I'm seeing is that the following pattern happens:

```
name: <extension_name>
typed_config:
  @type_url: <unrelated proto type URL, e.g. struct>
```

Requiring typed config does not help with this problem. This PR does help. I agree with the second step to deprecate missing typed_config, but that's a different problem, and we can do it in a followup.

Reply:
If an extension is being configured with the wrong type_url, that definitely seems like something that we should be rejecting.  But I think we could eliminate this through basically the same process I described above.  Basically, we would implement the warning message and the runtime option *after* we fail the type_url lookup, and it would be the same regardless of whether the type_url lookup fails because there was no type_url or because the type_url was invalid.

In other words, the code here would be something like this:

```
Factory* factory = Utility::getFactoryByType<Factory>(message.typed_config());
if (factory != nullptr) {
  return factory;
}
LogWarningMessage();
if (runtime_flag_allows_invalid_type_urls) {
  return Utility::getFactoryByName<Factory>(message.name());
}
return nullptr;
```

Reply:
Sure, we can combine the two changes in the same PR and re-use the runtime guard. Not sure about the warning message - would that be too noisy?

Reply:
All of Envoy's in-built extensions should have unique config types and hence distinct type URLs. This has been the case since the v3 cutover. Where are you seeing this pattern @kyessenov?

Reply:
All the distributed extensions follow this pattern. I've seen this in tests a lot and some external extensions.

Reply:
@htuch Is there a way to use a codepath for-by-name lookup only in existing tests? I try to get a sense of how much work it is to define protos for every test extension and that seems overwhelming.

Reply:
I'm on a journey to set runtime guard for all legacy tests using by-name lookup. I found internal use for `raw_buffer` inside listener manager which I fixed, too.

Reply:
@kyessenov I think we should file a bug for cleaning up the test extensions. It's a good bug for folks who want to make some early contributions I think (once we have an example PR to point them at to show what needs doing, which would be really useful).

Reply:
https://github.com/envoyproxy/envoy/issues/20476. The bulk of the work is done in this PR, we can make a small follow-up to demonstrate the clean-up.

Reply:
SGTM, thanks, I'm good with  this PR.